File: LSR_example.m
%This function was written by Ben Taylor @bentaylordata to demonstrate two
%approaches to this problem (LSR, and EWLSR). The first regression ignores
%the time effects, the second weights them with a 0.8 lambda decay, think
%EWMA. Tweet me @bentaylordata with hashtag #utahdatacompetition with any
%questions. 

context_map=csvread('train_X.csv');             % Load train input (context matrix)
sim_yield_die_loss=csvread('train_Y.csv');      % Load train output
wafer_count=size(context_map,1);                % Get number of wafers

%Build regular LSR model +++++++++++++++++++++++++++++++++++++++++++++++++
beta_LSR=pinv(context_map'*context_map)*context_map'*sim_yield_die_loss;   %regular LSR model


% Now load validation set and score
context_map_val=csvread('val_X.csv');         %Load train input (context matrix)
sim_yield_val_die_loss=csvread('val_Y.csv');           %Load train output

%Evaluate model
yp_LSR=context_map_val*beta_LSR;                   %Predict regression

%Make sure to cap crazy results outside possible die >600 <0
yp_LSR(yp_LSR<0)=0;
yp_LSR(yp_LSR>600)=600;

%Special cost function, under prediction is 10x penalty
err_LSR=yp_LSR-sim_yield_val_die_loss;              %Calculate errors

%Custom residual sum
err_underpredict_LSR=abs(err_LSR(err_LSR<0))*10;        %10x penalty
err_overpredict_LSR=abs(err_LSR(err_LSR>=0));

overall_score_LSR=(sum(err_underpredict_LSR)+sum(err_overpredict_LSR))/wafer_count   

File: LSR_example.py
from numpy import genfromtxt
from sklearn import linear_model

# Load training data
X = genfromtxt('train_X.csv', delimiter=',')
Y = genfromtxt('train_Y.csv', delimiter=',')

# Create linear regression object
#regr = linear_model.LinearRegression() 648.56
regr = linear_model.Ridge(alpha=.5)


# Train the model using the training sets
regr.fit(X, Y)

# Load validation data
X_val = genfromtxt('val_X.csv', delimiter=',')
Y_val = genfromtxt('val_Y.csv', delimiter=',')

# Now predict validation output
Y_pred = regr.predict(X_val)

# Crop impossible values
Y_pred[Y_pred < 0] = 0
Y_pred[Y_pred > 600] = 600

# Calculate score
err_Y=Y_pred-Y_val
score = (sum(abs(err_Y[err_Y<0]))*10+sum(abs(err_Y[err_Y>=0])))/len(err_Y)

print score

